config/repository_resource.go

```diff
@@ -0,0 +1,321 @@
+		filePath = filepath.Join(agentconfig.AptRepoDir(), "osconfig_managed_%s.repo")
+		if gpgkey != "" {
+			keyContents, err := fetchGPGKey(gpgkey)
+			if err != nil {
+				return nil, fmt.Errorf("error fetching apt gpg key %q: %v", gpgkey, err)
+			}
+
+			r.managedRepository.Apt.GpgFileContents = keyContents
+			r.managedRepository.Apt.GpgChecksum = checksum(bytes.NewReader(keyContents))
+			r.managedRepository.Apt.GpgFilePath = filepath.Join(aptGPGDir, "osconfig_added_"+r.managedRepository.Apt.GpgChecksum+".gpg")
+		}
+
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_RepositoryResource_Goo:
+		if !packages.GooGetExists {
+			return nil, errors.New("cannot manage googet repository because googet does not exist on the system")
+		}
+		r.managedRepository.GooGet = &GooGetRepository{RepositoryResource: r.GetGoo()}
+		r.managedRepository.RepoFileContents = googetRepoContents(r.GetGoo())
+		filePath = filepath.Join(agentconfig.GooGetRepoDir(), "osconfig_managed_%s.repo")
+
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_RepositoryResource_Yum:
+		if !packages.YumExists {
+			return nil, errors.New("cannot manage yum repository because yum does not exist on the system")
+		}
+		r.managedRepository.Yum = &YumRepository{RepositoryResource: r.GetYum()}
+		r.managedRepository.RepoFileContents = yumRepoContents(r.GetYum())
+		filePath = filepath.Join(agentconfig.YumRepoDir(), "osconfig_managed_%s.repo")
+
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_RepositoryResource_Zypper:
+		if !packages.YumExists {
+			return nil, errors.New("cannot manage yum repository because yum does not exist on the system")
+		}
+		r.managedRepository.Zypper = &ZypperRepository{RepositoryResource: r.GetZypper()}
+		r.managedRepository.RepoFileContents = zypperRepoContents(r.GetZypper())
+		filePath = filepath.Join(agentconfig.ZypperRepoDir(), "osconfig_managed_%s.repo")
+	default:
+		return nil, fmt.Errorf("Repository field not set or references unknown repository type: %v", r.GetRepository())
+	}
+
+	r.managedRepository.RepoChecksum = checksum(bytes.NewReader(r.managedRepository.RepoFileContents))
+	r.managedRepository.RepoFilePath = fmt.Sprintf(filePath, r.managedRepository.RepoChecksum[:10])
+	return &ManagedResources{Repositories: []ManagedRepository{r.managedRepository}}, nil
+}
+
+func contentsMatch(path, chksum string) (bool, error) {
+	file, err := os.OpenFile(path, os.O_RDONLY, 0644)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	defer file.Close()
+
+	return chksum == checksum(file), nil
+}
+
+func (r *repositoryResource) checkState(ctx context.Context) (inDesiredState bool, err error) {
+	// Check APT gpg key if applicable.
+	if r.managedRepository.Apt != nil && r.managedRepository.Apt.GpgFileContents != nil {
+		match, err := contentsMatch(r.managedRepository.Apt.GpgFilePath, r.managedRepository.Apt.GpgChecksum)
+		if !match {
+			return false, nil
+		}
+		if err != nil {
```

Comment:
Should this error check happen before the check for match as an error also would return a mismatch?

Reply:
yep, fixed